lib/chef/knife/winrm_knife_base.rb

```diff
@@ -0,0 +1,316 @@
+              password: config[:winrm_password],
+              port: config[:winrm_port],
+              operation_timeout: resolve_winrm_session_timeout,
+              basic_auth_only: resolve_winrm_basic_auth,
+              disable_sspi: resolve_winrm_disable_sspi,
+              transport: resolve_winrm_transport,
+              no_ssl_peer_verification: resolve_no_ssl_peer_verification,
+              ssl_peer_fingerprint: resolve_ssl_peer_fingerprint,
+              shell: config[:winrm_shell],
+              codepage: config[:winrm_codepage],
+            }
+
+            if @session_opts[:user] && (not @session_opts[:password])
+              @session_opts[:password] = config[:winrm_password] = get_password
+            end
+
+            if @session_opts[:transport] == :kerberos
+              @session_opts.merge!(resolve_winrm_kerberos_options)
+            end
+
+            @session_opts[:ca_trust_path] = config[:ca_trust_file] if config[:ca_trust_file]
+          end
+
+          def resolve_winrm_user
+            user = config[:winrm_user]
+
+            # Prefixing with '.\' when using negotiate
+            # to auth user against local machine domain
+            if resolve_winrm_basic_auth ||
+                resolve_winrm_transport == :kerberos ||
+                user.include?("\\") ||
+                user.include?("@")
+              user
+            else
+              ".\\#{user}"
+            end
+          end
+
+          def resolve_winrm_session_timeout
+            # 30 min (Default) OperationTimeout for long bootstraps fix for KNIFE_WINDOWS-8
+            config[:session_timeout].to_i * 60 if config[:session_timeout]
+          end
+
+          def resolve_winrm_basic_auth
+            config[:winrm_authentication_protocol] == "basic"
+          end
+
+          def resolve_winrm_kerberos_options
+            kerberos_opts = {}
+            kerberos_opts[:keytab] = config[:kerberos_keytab_file] if config[:kerberos_keytab_file]
+            kerberos_opts[:realm] = config[:kerberos_realm] if config[:kerberos_realm]
+            kerberos_opts[:service] = config[:kerberos_service] if config[:kerberos_service]
+            kerberos_opts
+          end
+
+          def resolve_winrm_transport
+            transport = config[:winrm_transport].to_sym
+            if config.any? { |k, v| k.to_s =~ /kerberos/ && !v.nil? }
+              transport = :kerberos
+            elsif transport != :ssl && negotiate_auth?
+              transport = :negotiate
+            end
+
+            transport
+          end
+
+          def resolve_no_ssl_peer_verification
+            config[:ca_trust_file].nil? && config[:winrm_ssl_verify_mode] == :verify_none && resolve_winrm_transport == :ssl
+          end
+
+          def resolve_ssl_peer_fingerprint
+            config[:ssl_peer_fingerprint]
+          end
+
+          def resolve_winrm_disable_sspi
+            resolve_winrm_transport != :negotiate
+          end
```

Comment:
This is used in one place, I'd probably inline this as well.

Reply:
Nuked